runtime/environment.go

```diff
@@ -0,0 +1,1147 @@
+		})
+		return
+	}
+}
+
+func (e *interpreterEnvironment) newCheckHandler() sema.CheckHandlerFunc {
+	return func(checker *sema.Checker, check func()) {
+		reportMetric(
+			check,
+			e.runtimeInterface,
+			func(metrics Metrics, duration time.Duration) {
+				metrics.ProgramChecked(checker.Location, duration)
+			},
+		)
+	}
+}
+
+func (e *interpreterEnvironment) newImportHandler(checkedImports importResolutionResults) sema.ImportHandlerFunc {
+	return func(
+		checker *sema.Checker,
+		importedLocation common.Location,
+		importRange ast.Range,
+	) (sema.Import, error) {
+
+		var elaboration *sema.Elaboration
+		switch importedLocation {
+		case stdlib.CryptoChecker.Location:
+			elaboration = stdlib.CryptoChecker.Elaboration
+
+		default:
+
+			// Check for cyclic imports
+			if checkedImports[importedLocation] {
+				return nil, &sema.CyclicImportsError{
+					Location: importedLocation,
+					Range:    importRange,
+				}
+			} else {
+				checkedImports[importedLocation] = true
+				defer delete(checkedImports, importedLocation)
+			}
+
+			program, err := e.getProgram(importedLocation, checkedImports)
+			if err != nil {
+				return nil, err
+			}
+
+			elaboration = program.Elaboration
+		}
+
+		return sema.ElaborationImport{
+			Elaboration: elaboration,
+		}, nil
+	}
+}
+
+func (e *interpreterEnvironment) GetProgram(location Location) (*interpreter.Program, error) {
+	return e.getProgram(location, importResolutionResults{})
+}
+
+// getProgram returns the existing program at the given location, if available.
+// If it is not available, it loads the code, and then parses and checks it.
+//
+func (e *interpreterEnvironment) getProgram(
+	location Location,
+	checkedImports importResolutionResults,
+) (
+	program *interpreter.Program,
+	err error,
+) {
+	wrapPanic(func() {
+		program, err = e.runtimeInterface.GetProgram(location)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if program == nil {
+		var code []byte
+		code, err = e.getCode(location)
+		if err != nil {
+			return nil, err
+		}
+
+		program, err = e.parseAndCheckProgram(
+			code,
+			location,
+			true,
+			checkedImports,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	e.codesAndPrograms.setProgram(location, program.Program)
+
+	return program, nil
+}
+
+func (e *interpreterEnvironment) getCode(location common.Location) (code []byte, err error) {
+	if addressLocation, ok := location.(common.AddressLocation); ok {
+		wrapPanic(func() {
+			code, err = e.runtimeInterface.GetAccountContractCode(
+				addressLocation.Address,
+				addressLocation.Name,
+			)
+		})
+	} else {
+		wrapPanic(func() {
+			code, err = e.runtimeInterface.GetCode(location)
+		})
+	}
+	return
+}
+
+func (e *interpreterEnvironment) newInterpreter(
+	location common.Location,
+	program *interpreter.Program,
+) (*interpreter.Interpreter, error) {
+
+	publicKeyValidationHandler := e.newPublicKeyValidationHandler()
+
+	options := []interpreter.Option{
```

Comment:
In a follow-up PR we can now make this a configuration struct that can be reused for all interpreters: All options only depend on the environment.